i#4678: DrStateCmp, a library for machine state comparison.

sapostolakis opened the pull request:
Implements drstatecmp, a library for machine state comparison for detecting instrumentation-induced clobbering of application state. This implementation supports checks for side-effect-free basic blocks (with the exception of the terminating instruction that could be a function call with side effects). The check involves executing two versions of these basic blocks, one with instrumentation and one without, and then comparing the resulting state to detect bugs. When a mismatch is detected, by default, the disassembled basic block that involved the bug will be emitted.

Introduces a new fifth instrumentation phase in drmgr (called post-instrumentation) to enable the drstatecmp library to perform the  duplication and the state comparison after all other instrumentation passes and without breaking any drmgr instrumentation contract (e.g., by adding app instructions in the instru2instru phase).

Illustrates the use of the drstatecmp library by providing a sample client.  This sample client introduces an instrumentation bug that is caught by drstatecmp. This sample client also shows how to specify an user-defined  callback to be invoked on state comparison mismatches.

Adds a fuzzing app for AArch64 that automatically generates from the decoding files a random sequence of side-effect-free and non-branch instructions with fuzzed operands for stress-testing clients with drstatecmp. This enables automatic testing of even newly inserted instructions in the decoder. 

Adds an option in drcachesim to enable drstatecmp and an AArch64 test that enables this option along with delayed tracing for the fuzzing app. Testing with an older DR version showed that this test would have eventually automatically detected #5023  (instrumentation-clobbering of aflags). 

Split up into 4 smaller PRs: #5059, #5090, #5107, #5125.

Issue: #4678

johnfxgalea commented:
I had a quick look at the PR! Thank you.

Not sure, but I think drbbdup can help you with your the bb copying, and facilitate the support of x86 32.

derekbruening commented:
> Not sure, but I think drbbdup can help you with your bb copying, and facilitate the support of x86 32.

I believe I mentioned drbbdup up front -- @sapostolakis you had reasons why it wouldn't work well in this case?  May be worth comments in the code on that front as well for future readers who see a multi-copy block and might wonder "why not drbbdup".

sapostolakis commented:
> > Not sure, but I think drbbdup can help you with your bb copying, and facilitate the support of x86 32.
> 
> I believe I mentioned drbbdup up front -- @sapostolakis you had reasons why it wouldn't work well in this case? May be worth comments in the code on that front as well for future readers who see a multi-copy block and might wonder "why not drbbdup".

Yep, I looked into drbbdup. It was well documented and it was very useful to understand how some things worked and get some ideas about duplication. I ended up not using it since just the duplication of the basic block is relatively simple (and a small part of the code, especially once more complete checking is added) and the use case is different enough that it cannot easily be used as it is (or with minimum changes).  In particular, drbbdup dispatches control according to runtime conditions, so one case is executed every time. Instead, I want to execute both the original basic block and the copy basic block. Besides, the biggest complexity of drstatecmp is the state comparison parts, and I do not have use for much of drbbdup's functionality. Finally, I also need to have ARM support so that would require extending and testing drbbdup for that as well.

For the x86-32 the issue is in the comparison part which is done in a clean call. eax seems to mismatch. Apart from the state comparison part, all the rest of the code is currently arch agnostic.

derekbruening commented on api/samples/statecmp.c:
```diff
@@ -0,0 +1,55 @@
+ */
+
+/* Illustrates using the drstatecmp extension.
+ *
```
This is not finished, yet the code doesn't say that: generally it's best to have TODO or other comments to make that clear.

What do you envision having here?  This should have some sample instrumentation to be tested or sthg.

I'm wondering whether this doesn't fit into api/samples/ which is for simple example tools, while drstatecmp is for comprehensive, sophisticated testing of instrumentation libraries and large tools.

derekbruening commented on ext/drstatecmp/drstatecmp.dox:
```diff
@@ -0,0 +1,146 @@
+
+The side-effect-free basic blocks are duplicated at drmgr's app2app stage. The copy of
+each such basic block is appended immediately after the original basic block. Labels are
+inserted in the beginning of the original basic block (ORIG_BB) and its copy (COPY_BB) to
```
Use `\p ORIG_BB` or sthg to indicate it's a literal code term.  Ditto for the others.

derekbruening commented on ext/drstatecmp/drstatecmp.dox:
```diff
@@ -0,0 +1,146 @@
+block is instrumented. In the beginning of the original basic block, the machine state is
+saved. This state will be used for state restoration before the re-execution of the basic
+block. At the end of the original block, code is inserted to save the state for later
+comparison and restore the machine state to the state before executing this bb. This stage
```
s/this bb/the other copy/ or sthg

derekbruening commented on ext/drstatecmp/drstatecmp.dox:
```diff
@@ -0,0 +1,146 @@
+saved. This state will be used for state restoration before the re-execution of the basic
+block. At the end of the original block, code is inserted to save the state for later
+comparison and restore the machine state to the state before executing this bb. This stage
+also removes the previously inserted jump instr to the TERM label and let it fall through
```
grammar: s/let/lets/

derekbruening replied:
Not understanding the inserted-and-then-removed jump instr: this is all static operation so why insert at all if it will jsut be removed?

derekbruening commented on ext/drstatecmp/drstatecmp.dox:
```diff
@@ -0,0 +1,146 @@
+to the instructions of the copy of the basic block. Then, the copy of the basic block is
+instrumented. At the end of the copy, a clean call is inserted that compares the current
+machine state with the machine state that was saved at the end of the original basic
+block. Finally, all previously-inserted (from other passes) instrumentation instructions
```
You mean from other app2app actions like rep string expansion?  How do you know how to undo such expansion?  It seems like you would instead need to have a high priority app2app callback to copy before anyone changed the code?

derekbruening commented on ext/drstatecmp/drstatecmp.dox:
```diff
@@ -0,0 +1,146 @@
+extra complexity to monitor and restore memory state). In this case, the checks are
+inserted throughout the basic block whenever needed taking into account the app
+instructions, existing instrumentation, and the lazy restoration by drreg. As a result,
+incomplete-decoder bugs and lazy-condition-ddreg bugs will not be detected.
```
"ddreg"

derekbruening commented on ext/drstatecmp/drstatecmp.dox:
```diff
@@ -0,0 +1,146 @@
+
+Instructions with side effects include instructions that write to memory, interrupts,
+and system calls. To avoid inter-procedural analysis, all function calls are
+conservatively assumed to have side effects.
```
I think this last sentence can be removed (function calls are are not embedded; they'll become separate blocks for each piece)

derekbruening commented on ext/drstatecmp/drstatecmp.dox:
```diff
@@ -0,0 +1,146 @@
+whether a basic block is side-effect free or not. The analysis stage is not used except
+for passing over the user_data to the insert instrumentation stage. In the latter stage,
+the user_data is also used to track when the instrumentation moves from the original
+basic block to its copy.
```
Hmm, this is more of an internal implementation detail.  I would expect these docs to focus more on how to use this library: like how to create a test from an existing complex client.  I want to create a version that includes these extra state comparisons for my test.  Do I have to use ifdefs to do two separate builds to avoid the checks in my release version?  Or is there a convenient flag where drstatecmp can be disabled without having to be ifdef-ed out?  If I link with it but place the drstatecmp_init call under a test-only flag in my client: will that do it, or will there still be overhead?

What apps should be executed for the test?  Up to the user presumably.

How are comparison failures reported?  Does the user get to decide whether the process is killed or whether there's just a printout or a callback to some user-controlled action?

derekbruening commented on ext/drstatecmp/drstatecmp.dox:
```diff
@@ -0,0 +1,146 @@
+\section sec_drstatecmp_machine_state_saved Machine States Saved
+
+During execution, thread-local storage is used to keep two machine states for each
+side-effect basic block. One is the state of the original basic block before its
```
side-effect-free?

derekbruening commented on ext/drstatecmp/drstatecmp.h:
```diff
@@ -0,0 +1,99 @@
+
+/**
+ * Priorities of drmgr instrumentation passes used by drstatecmp.  Users
+ * of drstatecmp can use the name DRMGR_PRIORITY_NAME_DRSTATECMP in the
```
Use `#` for a doxygen link: `#DRMGR_PRIORITY_NAME_DRSTATECMP`

derekbruening commented on ext/drstatecmp/drstatecmp.h:
```diff
@@ -0,0 +1,99 @@
+/**
+ * Priorities of drmgr instrumentation passes used by drstatecmp.  Users
+ * of drstatecmp can use the name DRMGR_PRIORITY_NAME_DRSTATECMP in the
+ * drmgr_priority_t.before field or can use these numeric priorities
```
Ditto: `#drmgr_priority_t.`

derekbruening commented on ext/drstatecmp/drstatecmp.h:
```diff
@@ -0,0 +1,99 @@
+    DRSTATECMP_SUCCESS,                   /* Operation succeeded. */
+    DRSTATECMP_ERROR,                     /* Operation failed. */
+    DRSTATECMP_ERROR_ALREADY_INITIALISED, /* DRSTATECMP can only be initialised once. */
+    DRSTATECMP_ERROR_NOT_INITIALIZED,     /* Operation failed: not initialized. */
```
Be consistent: you have INITIALISED on one line (British spelling) and INITIALIZED on the next, in the code and the comments.

derekbruening commented on ext/drstatecmp/drstatecmp.h:
```diff
@@ -0,0 +1,99 @@
+
+/** Success code for each drstatecmp operation. */
+typedef enum {
+    DRSTATECMP_SUCCESS,                   /* Operation succeeded. */
```
These should all be Doxygen comments `/**< */`

derekbruening commented on ext/drstatecmp/drstatecmp.h:
```diff
@@ -0,0 +1,99 @@
+typedef enum {
+    DRSTATECMP_SUCCESS,                   /* Operation succeeded. */
+    DRSTATECMP_ERROR,                     /* Operation failed. */
+    DRSTATECMP_ERROR_ALREADY_INITIALISED, /* DRSTATECMP can only be initialised once. */
```
nit: why all caps in the comment when no other docs use all caps?

sapostolakis replied:
Fixed.

derekbruening reviewed (commented):
Didn't look at drstatecmp.c yet: comments on the rest of the files.

derekbruening commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,572 @@
+     */
+    else if (match_label_val(instr, DRSTATECMP_LABEL_COPY_BB)) {
+        data->orig_bb_mode = false;
+        instr_t *prev_instr = instr_get_prev_app(instr);
```
A sanity check that prev_instr is not NULL and is what you expect would be good

sapostolakis replied:
Deleted code.

derekbruening commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,572 @@
+typedef struct {
+    dr_mcontext_t saved_state_for_restore; /* Last saved machine state for restoration. */
+    dr_mcontext_t saved_state_for_cmp;     /* Last saved machine state for comparison. */
+} drstatecmp_saved_states;
```
style: `_t` suffix for type names

sapostolakis replied:
Fixed.

derekbruening commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,572 @@
+    bool side_effect_free; /* Denotes whether the current bb has side-effects. */
+    bool orig_bb_mode; /* True when processing the instructions of the original version of
+                          a side-effect-free bb. Becomes false when starting processing
+                           the copies of these instructions*/
```
style: * on each line

style:  missing end `.`

sapostolakis replied:
Deleted code.

derekbruening commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,572 @@
+                         drstatecmp_label_t label_type, bool preinsert)
+{
+    instr_t *label = INSTR_CREATE_label(drcontext);
+    instr_set_meta(label);
```
instrlist_meta_preinsert does this, so this is redundant

derekbruening commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,572 @@
+
+/* Create and insert bb labels. */
+static instr_t *
+drstatecmp_insert_labels(void *drcontext, instrlist_t *ilist, instr_t *where,
```
Remove plural from name since it inserts *one* label?

sapostolakis replied:
Fixed.

derekbruening commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,572 @@
+{
+    /* Instructions with side effects include instructions that write to memory,
+     * interrupts, and syscalls. To avoid inter-procedural analysis, all function calls
+     * are conservatively assumed to have side effects.
```
The function call comment can be removed.

sapostolakis replied:
Removed.

derekbruening commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,572 @@
+     * are conservatively assumed to have side effects.
+     */
+    return instr_writes_memory(instr) || instr_is_interrupt(instr) ||
+        instr_is_syscall(instr) || instr_is_call(instr);
```
The instr_is_call can be removed: it should be just like any other branch for the purposes here

sapostolakis replied:
Removed.

derekbruening commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,572 @@
+    bool side_effect_free = true;
+    for (instr_t *inst = instrlist_first_app(bb); inst != NULL;
+         inst = instr_get_next_app(inst)) {
+        side_effect_free &= !drstatecmp_may_have_side_effects_instr(inst);
```
Avoid bitwise operators for bools, especially in C where there's no true bool type: safest to use logical &&

sapostolakis replied:
Done.

derekbruening commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,572 @@
+     */
+
+    /* Create a clone of bb. */
+    instrlist_t *copy_bb = instrlist_clone(drcontext, bb);
```
If sthg like rep str or scatter expansion went first and inserted emulation labels with stored data, cloning becomes a problem as the clone has the labels but they are missing some data (IIRC the data is set to NULL).  Looks like the late priority will have this go after those expansions.  Better to go first?

derekbruening replied:
Hmm, I was thinking it would be good to have the original rather than expanded version (i.e., single scatter instruction; single rep movs insruction), so going first was always good.  But for true emulation where the original won't even run on the underlying hardware: you'd want to go after that.

derekbruening replied:
Actually for the rep movs I guess you couldn't take the original which would loop all N iters at once and compare to the expanded version which would do just one iter for each state cmp.

sapostolakis replied:
Handle app2app transformations with new implementation.

derekbruening commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,572 @@
+    opnd_t term_label_opnd = opnd_create_instr(term_label);
+
+    /* Replace the terminating instruction of the original bb with a jmp to the
+     * terminating instruction of the bb clone (if any).
```
What is the purpose of this added jump?  Other instrumentation libraries/clients are not likely to look at it and conclude that it skips the copy and then ignore the copy: they'll assume linear control flow and blindly instrument every instruction in the list, ignoring the fact that there's a jump here.

johnfxgalea replied:
@derekbruening Without the jmp, one can have tools, like drreg, which will consider the orig and copy blocks as one whole basic block and therefore perform incorrect analysis (for drreg's case liveness). Think of a situation where a register used in the orig block is erroneously considered dead only because the copy block writes to it further down the line.

Edit: I just read that you raised a similar point further down in your review.

sapostolakis replied:
The jump is not used anymore.

derekbruening commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,572 @@
+        instrlist_postinsert(bb, term_inst, jmp_term);
+    }
+
+    /* Append the instructions of the bb copy to the original bb. */
```
But this will present the copy to all the instrumentation libraries and clients.  So drreg will consider all the reg uses in the copy when deciding which scratch regs to use, which will warp its behavior so it won't match what it would do w/o a copy.  IIRC its liveness analysis does consider jumps?  Is that why a jump was added?  Please add comments if so.  But its reg use counts will ignore the jump.  Plus other libraries/clients frequently count app instrs or take actions across app instrs: the most common is putting a clean call or batched action at the end of the block, which would mean inside this copy.  Then below you remove all the instrumentation: so the resulting block will not contain the instrumentation in the original part that would have been there.

I think this copy has to be hidden from the instrumentation, and only added in the insru2instru phase or something like that.

derekbruening replied:
@johnfxgalea -- Any thoughts on the best way to hide the extra copy from instrumentation?  It seems like either using the drbbdup approach with a drmgr-involved modified interface to hide the other copies, or adding it in the instru2instru phase (or I suppose insertion w/ very low priority) are the options: with the latter technically breaking rules on adding app instructions.

sapostolakis replied:
Yep, the jump was added to allow drreg to do liveness analysis correctly but you are right that in many cases linear control flow is assumed and the copy will affect various instrumentation passes. To me doing everything in the instru2instru phase seems to be the best choice. Besides the main motivation for the comparison library is to check for clobbering by instrumentation. So, the safest option is to add the checks at the very end without influencing any other instrumentation. The code will also be simpler, since there is bb-level scope in instru2instru and everything will done in a single pass.

johnfxgalea replied:
instru2instru does seem simple, but I don't really like the idea of breaking drmgr's instrumentation contract with the addition of app instructions, especially if this becomes a common occurrence in the future. Then again, other solutions require more work.

If performance is not a major problem, an alternative approach might be to rely on re-direction, i.e., with the use of drbbdup,  first execute the instrumented case. Then rely on a clean call, which is placed at the end, to set a flag and redirect execution again to the same basic block. The active flag will instead lead drbbdup to dispatch to the uninstrumented case (which will only include instrumentation to reset the flag).

sapostolakis replied:
Performance is not that critical for now (although it might be in the future if it proves useful and it is used beyond test apps), but code simplicity is. The point of this lib is to detect bugs in other DR components while it remains bug-free itself. If it is too complex or relies on other extensions (and thus inherit their complexities and potential bugs) then its usefulness decreases. Bugs in drstatecmp or its deps could lead to time-wasting false positives (incorrect detection of mismatch that needs to be investigated manually) or false negatives (undetected clobbering). I would also count as detrimental any intrusive effect of drstatecmp or its deps on the behavior of other DR components.

johnfxgalea replied:
All great points, and a big fan of simplicity.

> while it remains bug-free itself

The only issue I see is that going against the specification of drmgr is a bug in itself.

We could document the violating use of the instru2instru phase really well, stressing that DrStateCmp is just a debugging tool, but if we can find a better way to avoid this exception to the rule, that would be great! In general, exceptions to rules could potentially lead to nasty bugs or maintenance issues in the future, and this is something we should avoid.

Keeping in mind points on simplicity, how about we introduce a new final special drmgr phase (maybe call it `bb2debug`?), which is suited for DrStateCmp and other debugging extensions that we might create in the future? The phase will be the the last phase considered by drmgr and allows for flexible instrumentation, including app instr. Its introduction appears that it won't break drmgr rules.

sapostolakis replied:
Having a fifth drmgr phase might be a good idea. The fourth phase is for instrumentation optimizations and there will be a fifth phase that is for instrumentation debugging. Seems appropriate. I am fine with more work as long as DrStateCmp itself remains simple.

johnfxgalea replied:
Yeah! If we do include the fifth stage, I'm not sure if we should document it solely for use by so called "debugging" extensions or allow other "normal" extensions to possibly register it as well. There could be some small differences in the implementation. If we go for phase 5 to be strictly for debugging tools, then we could potentially have an internal header file, and leave drmgr.h unchanged.

Maybe @derekbruening also has an opinion on this?

derekbruening replied:
My first reaction is to be hesitant to change the core phases: I started implementing a split of the insertion phase and ended up bailing because of complexities with drbbdup and other existing users of the current phases (filed aborted attempt as #4947).  But a completely separate phase that doesn't touch the existing key phases of app2app-analysis-insertion is probably going to have minimal disruption so it seems fine in that respect.

I kind of like the drbbdup re-execute idea at a high level: drbbdup is an existing framework supporting construction of two separate versions, which is exactly what we want here.  But having to do the full redirect to chain two versions, which drbbdup wasn't designed for, makes it much less appealing.

> If we go for phase 5 to be strictly for debugging tools, then we could potentially have an internal header file, and leave drmgr.h unchanged.

Could you elaborate on this?  I assumed we would want a public phase 5 since we want these debugging tools to be first-class libraries usable with new instrumentation schemes.

johnfxgalea replied:
Put simply, we need to decide whether we should have access to phase 5 via drmgr_priv.h or drmgr.h. If we decide that phase 5 is solely intended for debugging tools, maybe drmgr_priv.h would be sufficient. If we envisage phase 5 to be used more generally, then drmgr.h ?

derekbruening replied:
OK I could see separating, or documenting inside drmgr.h that it is meant for only certain uses.  But even if separated it would still be a published header using exported drmgr library functions, right?  So it wouldn't really be private/internal: just documented as for narrow use cases.

johnfxgalea replied:
But even if separated it would still be a published header using exported drmgr library functions, right? 

Yep, but generally users should include drmgr.h and not drmgr_priv.h.

derekbruening replied:
OK, though as a nit I would have a different name, since _private.h in DR means an interface not exported at all outside a module/library.

derekbruening reviewed (commented):
Finishing off the initial review.

johnfxgalea commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,572 @@
+    dr_get_mcontext(drcontext, &mc_expected);
+
+#ifdef X86
+    drstatecmp_check_gpr_value("xdi", mc_instrumented->xdi, mc_expected.xdi);
```
I would add all these debug checks to a helper function.

johnfxgalea commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,572 @@
+    else if (match_label_val(instr, DRSTATECMP_LABEL_COPY_BB)) {
+        data->orig_bb_mode = false;
+        instr_t *prev_instr = instr_get_prev_app(instr);
+        instrlist_remove(bb, prev_instr);
```
If I am not mistaken, you cannot remove app instructions (inserted at app2app) during the insertion phase. This breaks DR's contract.

Moreover, there could have been an application code analysis pass (so between app2app and insertion) that takes into account the jmp.

sapostolakis replied:
Addressed by moving all changes to the new final phase.

johnfxgalea reviewed (commented):
My major concern is the removal of the jmp instr during the instrumentation phase.

sapostolakis commented:
Added a 5th instrumentation phase in drmgr, as discussed, and updated drstatecmp to use it. 
Converted this PR to a draft since, before further reviewing, I need to refine it a bit, address all previous review comments, and update the documentation regarding the newly introduced instrumentation phase.

sapostolakis commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,572 @@
+    }
+
+    /* Free the user_data when we are done with this bb. */
+    if (instr == instrlist_last(bb))
```
Deleted code.

sapostolakis commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,572 @@
+     */
+    if (!instr_is_app(instr)) {
+        instrlist_remove(bb, instr);
+        instr_destroy(drcontext, instr);
```
Fixed with new implementation.

sapostolakis commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,572 @@
+    }
+
+    /* Remove all instrumentation code in the bb copy to detect any
+     * unintended clobbering by instrumentation.
```
Fixed.

derekbruening commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,606 @@
+        if (drmgr_is_emulation_start(inst)) {
+            emulated_instr_t emulated_inst;
+            drmgr_get_emulated_instr_data(inst, &emulated_inst);
+            if (!(emulated_inst.flags & DR_EMULATE_INSTR_ONLY)) {
```
We use `TEST`

sapostolakis replied:
Done.

derekbruening commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,606 @@
+ * Save a pre-app2app copy of side-effect-free basic blocks.
+ * It is assumed that this pass has the highest priority among all app2app passes
+ * and thus it is able to capture the pre-app2app state.
+ *
```
nit: empty line at unusual place

sapostolakis replied:
Removed.

derekbruening commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,606 @@
+                 * be the post-app2app copy.
+                 */
+                instrlist_clear_and_destroy(drcontext, data->pre_app2app_bb);
+                data->golden_bb_copy = instrlist_clone(drcontext, bb);
```
Best to have a test case for this path

sapostolakis replied:
Should be exercised now with the fuzz app.

derekbruening commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,606 @@
+     */
+
+    /* Get an instrumentation-free copy of the bb which is the golden copy kept
+     * in the user_data*/
```
style: missing space and newline

sapostolakis replied:
Done.

derekbruening commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,606 @@
+     */
+    instr_t *term_inst_copy_bb = instrlist_last_app(copy_bb);
+    bool preinsert = true;
+    if (!instr_is_cti(term_inst_copy_bb) && !instr_is_return(term_inst_copy_bb))
```
"cti" should include return

sapostolakis replied:
You are right. Double-checked it and indeed instr_is_cti() checks for instr_is_mbr_arch() which is a superset of instr_is_return().

derekbruening commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,606 @@
+    drstatecmp_compare_state(drcontext, bb, labels->term);
+}
+
+/* Duplicate the side-effect basic block for re-execution and add saving/restoring of
```
side-effect free

sapostolakis replied:
Done.

derekbruening reviewed (commented):
I think the only major issue is adding a test.

Not sure if @johnfxgalea wanted to also take a look?

johnfxgalea commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,606 @@
+drstatecmp_save_state(void *drcontext, instrlist_t *bb, instr_t *instr, bool for_cmp)
+{
+    dr_insert_clean_call(drcontext, bb, instr, (void *)drstatecmp_save_state_call,
+                         false /*fpstate */, 1, OPND_CREATE_INT32((int)for_cmp));
```
Nit: Add space for fpstate comment.

sapostolakis replied:
Done.

johnfxgalea commented:
Apart from @derekbruening 's comments, just one nit from my end.

sapostolakis commented on ext/drmgr/drmgr.h:
```diff
@@ -515,6 +518,42 @@ bool
+ * basic block.  drmgr will call \p func as the fifth of five
+ * instrumentation stages for each dynamic application basic block.
+ * Post-instrumentation passes are allowed to insert both meta and
+ * not non-meta instructions, and are intended for debugging prior
```
Fixed.

sapostolakis commented on ext/drstatecmp/drstatecmp.dox:
```diff
@@ -0,0 +1,136 @@
+The initialization routine \p drstatecmp_init() enables the insertion of the machine state
+checks. When drstatecmp is linked but the client does not invoke \p drstatecmp_init(),
+there are no inserted checks and no overhead to the client. Thus, a client can wrap the
+initialization of drstatecmp with a flag or a ifdef and invoke it only on debug builds.
```
Done.

sapostolakis commented on ext/drstatecmp/drstatecmp.dox:
```diff
@@ -0,0 +1,136 @@
+original app instructions requires emulation (i.e., not executable).
+
+The duplication of side-effect-free basic blocks and the checks insertion occurs at
+drmgr's post-instrumentation stage. A instrumentation-free copy of each side-effect-free
```
Done.

sapostolakis commented on ext/drstatecmp/drstatecmp.dox:
```diff
@@ -0,0 +1,136 @@
+
+The duplication of side-effect-free basic blocks and the checks insertion occurs at
+drmgr's post-instrumentation stage. A instrumentation-free copy of each side-effect-free
+basic block is appended immediately after the original/instrumented version of the basic
```
Done.

sapostolakis commented on ext/drstatecmp/drstatecmp.dox:
```diff
@@ -0,0 +1,136 @@
+The duplication of side-effect-free basic blocks and the checks insertion occurs at
+drmgr's post-instrumentation stage. A instrumentation-free copy of each side-effect-free
+basic block is appended immediately after the original/instrumented version of the basic
+block. Before the execution of the original basic block, its machine state is saved. This
```
Yes, just using instrumented is clearer.

sapostolakis commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,606 @@
+ * DAMAGE.
+ */
+
+#include "dr_defines.h"
```
Removed it.

sapostolakis reviewed (commented):
Fixed remaining issues except for adding a test outside the sample client. Will add tomorrow.

sapostolakis commented:
Actually code_api|sample.statecmp failed on AArch64, so I will need to fix the sample client for AArch64 as well.

sapostolakis commented:
The AArch64 error and testing with drreg revealed some corner cases leading to false positives. drreg will not bother to spill (and later restore) a dead register. So, instrumentation is allowed to clobber without care a register that will be written by an app instr later in the basic block without being read first. This is mostly okay for drstatecmp since at the end of the basic block the register will contain the written app value and the checks will work out. However, if the app instr that overwrites the register is the terminating instruction of the basic block, then it is not re-executed and precedes the checks, leading to a "false" positive. It is not really a false alarm since there is in fact clobbering but it is expected and not harmful so this alarm is undesirable. In the AArch64 example, the terminating example was a 'blr' instr which sets the link register (x30), but this register was used as a scratch register and it was never spilled/restored by drreg since it will be overwritten at the end of the basic block without being read before that by any other app instr.

sapostolakis commented:
Another source of false positives are side-effect-free instructions that yield on re-executions different results with the same register state input. One such instruction is rdtsc. A re-execution will lead to a different result triggering a false alarm. Could be seen as "non-deterministic" results. Sadly, x86 should have a lot of opcodes with the same issue.

sapostolakis commented:
Added a comprehensive fuzzing app for AArch64 that automatically generates from the decoding files (codec.txt) a random sequence of side-effect-free and non-branch instructions with fuzzed operands for stress-testing clients with drstatecmp. 

A test that uses this fuzzing app with delayed tracing and enabled drstatecmp has also been added for drcachesim.

By default the app generates a basic block with 10k random instructions. 9-10% of these instructions  (while being decodable) end up leading to SIGILL and are thus skipped and 1.5-2% of produced encoding cannot be decoded. Therefore around 9k instructions are eventually used each time. To avoid flakiness, I added a handle for SIGSEGV and SIGFPE although it should be rare to see such errors, especially after limiting the Rd to 0-15 gpr.
The test takes 7sec for release and 25sec for a debug AArch64 build.

sapostolakis commented:
Adding @AssadHashmi as a reviewer since this PR involves changes in codec.py/c and is related to decoding efforts and improved AArch64 testing. Updated the PR description to provide more context.

sapostolakis commented on suite/tests/client-interface/drx-scattergather.dll.c:
```diff
@@ -330,7 +330,7 @@ dr_init(client_id_t id)
+    ok = drmgr_register_bb_instrumentation_ex_event(
+        event_bb_app2app, event_bb_analysis, event_app_instruction, NULL, NULL, NULL);
     CHECK(ok, "drmgr register bb failed");
 }
```
It now has two tests. One with the sample client for all archs and one more comprehensive for AArch64 with the fuzzing app and drcachesim.

AssadHashmi commented on core/ir/aarch64/codec.py:
```diff
@@ -338,6 +338,39 @@ def generate_opcode_names(patterns):
+    # iii) SVE memory (opcode: 0010 and 1 for the most significant bit).
+    # The allowed instructions include all the Data Processing instructions (including
+    # the FP and SIMD ones) and the non-memory SVE instructions.
+    excluded_opcodes = re.compile('....1.0|...101.|1..0010')
```
AIUI the MSB is only 1 for the SME (Scalable Matrix Extension) instructions.

sapostolakis replied:
Thanks for taking a look.
For finding the opcodes to exclude I relied on the Armv8 documentation of the encodings (https://developer.arm.com/documentation/ddi0596/2021-06/Index-by-Encoding/SVE-encodings?lang=en). In this documentation, SVE encodings with MSB 1 are marked as SVE memory. I assume that the newly introduced SME in Armv9 would also involve memory ops. If there are non-memory-involving ops in there that could be included in the fuzz app, we can always relax this exclusion in the future.

AssadHashmi replied:
You are right, my bad, I was looking at old documentation :-/

derekbruening commented:
This is an extremely large diff (making it hard to review (more likely for review to miss pieces, burden on reviewer to find large chunk of time, etc.), harder to revert just pieces of it if sthg goes wrong: just harder to deal with): `git diff origin/master..` is 4751 lines!   We consider 1500 lines to be large: https://dynamorio.org/page_code_reviews.html#autotoc_md110

Ordinarily I would request that it be split up and would not even attempt to review it as one: the drcachesim portion should be a separate PR; the drmgr 5th stage could go in first as its own PR w/o any drstatecmp code.  Given the circumstances though I'm making my way through it; part-way at the moment.

sapostolakis commented:
> Ordinarily I would request that it be split up and would not even attempt to review it as one: the drcachesim portion should be a separate PR; the drmgr 5th stage could go in first as its own PR w/o any drstatecmp code. Given the circumstances though I'm making my way through it; part-way at the moment.

I think it is around 1.8K lines but I agree it is too big and I appreciate you taking the time to review it anyway. It has been reviewed incrementally and older parts have not changed but it has been a long time since then. As changes were requested, I kept updating this PR but I should have spawn new PRs for some of the changes, especially for the new drmgr stage that is quite separate from this library. Even if you get through this PR, I do not mind splitting it up for maintainability reasons (easier to revert part of it etc.).

derekbruening commented:
> I think it is around 1.8K lines 

FTR, I've always measured as the unified diff format line count: 

```
$ git checkout i4678-add-state-comp-lib
$ git diff origin/master.. | wc
   4751   34499  259491
```

But it looks like this branch is not merged w/ the latest master...so I think we want:
```
$ git diff $(git merge-base i4678-add-state-comp-lib origin/master).. | wc
   2451   11809  102941
```

sapostolakis commented:
From the github GUI I was seeing +1,735 −86  as this PR diff. But in any case, big PR. I will spawn some new PRs to make it easier to maintain (and review if you have not finished looking into this one).

I also still need to fix the windows failure. I think adding the 5th instrumentation stage led to the missing __chkstk for building drmgr. So, another reason to split it is to isolate the windows issue and resolve it.

derekbruening commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,691 @@
+ * golden copy for comparison. There are two options: i) pre-app2app-phase copy, where
+ * the code just contains the original app instructions, and ii) post-app2app-phase copy.
+ * The first option can catch bugs in app2app passes and it is selected unless any of
+ * original app instructions requires emulation (true emulation). In that case, the
```
nit: s/original/the original/

sapostolakis replied:
Done.

derekbruening commented on ext/drstatecmp/drstatecmp.c:
```diff
@@ -0,0 +1,691 @@
+        void *drcontext = dr_get_current_drcontext();
+        instrlist_t *bb = decode_as_bb(drcontext, bb_start_pc);
+        dr_fprintf(STDERR, "Application basic block where mismatch detected: \n");
+        instrlist_disassemble(drcontext, bb_start_pc, bb, STDERR);
```
Having the mcontext values would also be useful: how about a TODO or XXX comment for now.

sapostolakis replied:
Done.

derekbruening commented on ext/drstatecmp/drstatecmp.dox:
```diff
@@ -0,0 +1,137 @@
+instrumented basic block. At the end of the instrumentation-free copy of the basic block,
+a clean call is inserted that compares the current machine state of the
+instrumentation-free basic block with the machine state that was saved at the end of the
+instrumented basic block. An user-supplied callback is invoked in case of any state
```
nit: s/An user/A user/

sapostolakis replied:
Done.

derekbruening commented on suite/tests/CMakeLists.txt:
```diff
@@ -3302,6 +3307,11 @@ if (BUILD_CLIENTS)
     endif ()
 
+    # Stress-test drcachesim with drstatecmp checks.
+    if (AARCH64)
```
I know the fuzz app is a64-only -- but isn't this test available for x86?

sapostolakis replied:
drcachesim with drstatecmp can tested for x86 for a different app. 
I added another drcachesim test with drstatecmp for the the ci_shared_app (so available for all archs).
Btw the code_api|sample.statecmp test is available for all archs.

derekbruening commented on suite/tests/client-interface/drstatecmp-fuzz-app.c:
```diff
@@ -0,0 +1,218 @@
+ */
+
+/* Fuzzing application to stress-test DR with the drstatecmp library.
+ * Only AArch64 is currently supported.
```
Add ifdefs to `#error` if AARCH64 is not set

sapostolakis replied:
Done.

derekbruening commented on suite/tests/client-interface/drstatecmp-fuzz-app.c:
```diff
@@ -0,0 +1,218 @@
+append_instr(instr_t *instr, byte *encode_pc)
+{
+    print_instr_pc(instr, encode_pc);
+    byte *nxt_pc = instr_encode(GLOBAL_DCONTEXT, instr, encode_pc);
```
Check the return value to ensure it encoded successfully

sapostolakis replied:
Checked with assertion two lines down but moved it immediately after to make it clearer.

derekbruening commented on suite/tests/client-interface/drstatecmp-fuzz-app.c:
```diff
@@ -0,0 +1,218 @@
+static byte *
+append_instr(instr_t *instr, byte *encode_pc)
+{
+    print_instr_pc(instr, encode_pc);
```
nit: Better to avoid the call if verbose is turned off

sapostolakis replied:
Done.

derekbruening commented on suite/tests/client-interface/drstatecmp-fuzz-app.c:
```diff
@@ -0,0 +1,218 @@
+main(void)
+{
+    /* Produce fuzzing application code. */
+    srand(time(NULL));
```
Add an XXX comment about taking in the seed as a parameter, and printing it out on an error, to reproduce the exact behavior in a subsequent run.

sapostolakis replied:
Done.

derekbruening commented on suite/tests/client-interface/drstatecmp-fuzz-app.c:
```diff
@@ -0,0 +1,218 @@
+
+#include "configure.h"
+#include "dr_api.h"
+#include "opcode_opnd_pairs.h"
```
Maybe worth a comment here that this is a special file generated from the codec just for this fuzzer

sapostolakis replied:
Done.

derekbruening commented on suite/tests/client-interface/drstatecmp-fuzz-app.c:
```diff
@@ -0,0 +1,218 @@
+    const opcode_opnd_pair_t opcode_opnd_pair = fuzz_opcode_opnd_pairs[opcode_pick];
+
+    uint32_t opnd_mask = opcode_opnd_pair.opnd;
+    /* Avoid registers 16-31 for Rd (special registers and callee-saved registers). */
```
I think this needs a little more discussion: this reduces the space of errors that will be found.  If you could add a comment discussing the benefits and drawbacks and how you might handle these with additional state saving or sthg.

sapostolakis replied:
I added a comment on the reasoning of this constraint and how to relax it.

derekbruening reviewed (approved):
May be nice to split this into separate PR's as discussed.

sapostolakis commented on ext/drmgr/drmgr.h:
```diff
@@ -515,6 +515,79 @@ bool
+ * default priority is used.
+ */
+bool
+drmgr_register_bb_post_instru_ex_event(drmgr_app2app_ex_cb_t app2app_func,
```
Oh I didn't realize the naming convention there. Changed it to drmgr_register_bb_instrumentation_all_events() and ditto for unregister.

sapostolakis commented on core/ir/aarch64/codec.py:
```diff
@@ -338,6 +338,39 @@ def generate_opcode_names(patterns):
           '#endif /* OPCODE_NAMES_H */']
     return '\n'.join(c) + '\n'
 
+def generate_opcode_opnd_pairs(patterns):
```
Done.

sapostolakis commented on api/samples/statecmp.c:
```diff
@@ -0,0 +1,180 @@
+    /* To enable state comparison checks by the drstatecmp library, a client simply needs
+     * to initially invoke drstatecmp_init() and drstatecmp_exit() on exit.
+     * The invocation of drstatecmp_init() registers callbacks that insert
+     * machine state comparison checks in the code. An user-provided callback is
```
Done.

sapostolakis commented on api/samples/statecmp.c:
```diff
@@ -0,0 +1,180 @@
+}
+
+static bool
+bb_may_have_side_effects(void *drcontext, instrlist_t *bb)
```
Just to clarify, no other client will need to duplicate drstatecmp internal logic to use drstatecmp. This function does not specify where checks are enabled but it is just an artifact of creating a buggy client that does not lead to a flaky test. This client introduces clobbering of aflags with its instrumentation. To avoid a seg fault or any form of weird behavior, I had to restrain the client to only introduce bugs in basic blocks that drstatecmp could test.  
That said, it is worth exporting this functionality to avoid the redundancy here (and getting out-of-sync with the test client). I also renamed it to drstatecmp_bb_checks_enabled() for future-proofing as suggested.

sapostolakis commented on api/samples/statecmp.c:
```diff
@@ -0,0 +1,180 @@
+
+/* Illustrates use of the drstatecmp extension by a sample client.
+ * This sample client introduces an instrumentation bug that is caught by
+ * drstatecmp. This sample client also shows how to specify an user-defined
```
Done.

sapostolakis commented:
For x86 in DR an undefined effect on eflags is considered to write to eflags since programs shouldn't be reading eflags after an undefined modification to them. However, drstatecmp was checking the architectural state through re-execution and thus was catching clobbering of the flags by instrumentation in cases where drreg relied on undef behavior to consider the flags to be dead. Avoided checking the flags in cases with dead flags to avoid these "false" positives.

sapostolakis commented:
Split the new instrumentation phase to a new separate PR (https://github.com/DynamoRIO/dynamorio/pull/5059).

sapostolakis commented:
As discussed, this PR has been split up into 4 smaller PRs: #5059, #5090, #5107, #5125.
